### ISSUE TITLE
Using slc to start Loopback breaks promises?

### ISSUE THREAD

**Issue by esco** [strongloop/strongloop#250]
The error is as described in https://github.com/strongloop/loopback-datasource-juggler/issues/719 after investigation is seems that its triggered by using `slc run` to start the app. Using regular `node` solves the issue for me, but I wanted to create this issue for anyone who might experience the same thing and for future developments.

**Comment by sam-github** [comment]
The linked issue doesn't have any reproduction information. Could you point to a sample app, or describe how to build a sample app, that will exhibit the behaviour you see?

Also, you say "slc run", but you don't say whether you are running in clustered mode or not (what the full "slc run" command line is).

I can't think of anything supervisor does that could interact with promises, but clearly something is going wrong.

**Comment by sam-github** [comment]
@raymondfeng do you have any ideas what is happening?

**Comment by esco** [comment]
Hey @sam-github the linked issue does have reproduction information in the first comment, I'll re-paste it here for convenience 

OSX
node 0.12.7
strongloop@5.0.0

----
Relevant dependencies
``` json
{
    "loopback-connector-mongodb": "^1.13.0",
    "loopback-datasource-juggler": "^2.39.0"
}
```

To replicate:

```
$ git clone https://github.com/ForbesMobile/promise-failure
$ npm i -g strongloop
$ cd promise-failure
$ npm i
$ slc run
```
> https://github.com/ForbesMobile/promise-failure made by @senorcris

Visit http://127.0.0.1:3000/explorer/#!/People/test press the "Try it out!" button. You will get the following error:

```
{
  "error": {
    "name": "TypeError",
    "status": 500,
    "message": "Cannot read property 'then' of undefined",
    "stack": "TypeError: Cannot read property 'then' of undefined\n    at Function.Person.test (/Volumes/Dev/servers/node/test/promise-failure/common/models/person.js:5:7)\n    at SharedMethod.invoke (/Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/shared-method.js:248:25)\n    at HttpContext.invoke (/Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/http-context.js:384:12)\n    at /Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/remote-objects.js:620:11\n    at execStack (/Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/remote-objects.js:460:7)\n    at RemoteObjects.execHooks (/Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/remote-objects.js:464:10)\n    at /Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/node_modules/strong-remoting/lib/remote-objects.js:617:10\n    at /Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/lib/application.js:357:13\n    at /Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/lib/model.js:313:7\n    at /Volumes/Dev/servers/node/test/promise-failure/node_modules/loopback/common/models/acl.js:465:23"
  }
}
```

Remote method being called:
```javascript
Person.test = function(cb) {
    Person.find({
      name: "Jake"
    }).then(function(p) {
      console.log('p: ', p);
      cb(null, p);
    }).catch(function(err) {
      console.log(err);
      cb(err);
    });
  };
```

Person is a PersistedModel
```json
{
  "name": "Person",
  "base": "PersistedModel",
  "idInjection": true,
  "options": {
    "validateUpsert": true
  },
  "properties": {
    "name": {
      "type": "string"
    }
  },
  "validations": [],
  "relations": {},
  "acls": [],
  "methods": {}
}
```
Person.find should return a promise, but instead returns `undefined`

Any idea why PersistedModel.find isn't returning a promise?

**Comment by esco** [comment]
Also to answer your earlier question it happens in both non-cluster (locally) and cluster modes (on Heroku with NODE_ENV=production).

**Comment by sam-github** [comment]
Setting globals.Promise to bluebird strikes me as unsafe, you may want to reconsider that approach, Promise is native in recent node releases, so this will replace it with a not necessarily perfectly compatible version.

Your app doesn't install on node v3.x, strangely, bluebird doesn't get installed, I had to install it manually.

That aside, I can reproduce, it is related to strong-agent, running as `slc run --no-profile` fixes the issue. I'll look into it.

**Comment by esco** [comment]
Right, ideally it would be `global.Promise = global.Promise || require('bluebird')` but this was just a quick sample. The original version didn't have bluebird and still had the error, we talk about it a little in the linked issue. Thanks for looking into this so quickly and pointing out the `--no-profile` option :-).

**Comment by sam-github** [comment]
Of course, if you are using `--no-profile`, there is practically no reason to use `slc run` (well, it still clusters and restarts, so I guess some use).

**Comment by bnoordhuis** [comment]
@cgole Why did you assign this to me?

**Comment by cgole** [comment]
We are triaging bugs and @chandadharap  thought you should fix this one.

**Comment by clarkorz** [comment]
@sam-github I have a clue. This only happens when using strong-agent 2.x

```
cb = cb || utils.createPromiseCallback();
```
The DAO methods return a promise only if cb is not set. But when running with `slc run`, strong-agent will wrap DAO methods and pass a proxy as cb.

**Comment by bnoordhuis** [comment]
@clarkorz You're right, that's what happens.  There is a fix under way.

**Comment by poliver87** [comment]
In the mean time, is there a way to disable strong-agent when using strong-pm on a production host?

**Comment by glesage** [comment]
@bnoordhuis There is a fix under way? Anywhere we could see status or contribute?

Thanks!

**Comment by bnoordhuis** [comment]
@glesage There's a fix ready to be released but the release is pending some non-technical issues.  I'll check if we can do a patch release next week.

**Comment by glesage** [comment]
Ok that would be great! We've switched all our callbacks to promises on our codebase and I like to stay up to date with latest versions of strong-pm... (:

**Comment by Shyri** [comment]
Hi
Any update on this? Im having the same issue with strong-pm.
Thanks!

**Comment by danielrvt** [comment]
Same issue here... struggled for hours and hours... really annoying... Finally I'm just using PM2

**Comment by sam-github** [comment]
Fix has been published as strong-agent@2.0.2

**Comment by javefang** [comment]
Thank you for fixing this. Confirmed working now!
